File: GarbageV3-2025/app/src/main/java/dk/itu/GarbageV3/ItemsDB.java
package dk.itu.GarbageV3;

import android.content.Context;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class ItemsDB {
    private static ItemsDB sItemsDB;

    private final HashMap<String, String> ItemsDB= new HashMap<String, String>();

    private static Context context;

    private ItemsDB() throws IOException {
        if(context == null){
            throw new IllegalStateException("context must be set first!");
        }
        fillItemsDB(context, "garbage.txt");
    }

    // Singleton pattern get implementation
    public static ItemsDB get(){
        if (sItemsDB == null) sItemsDB = new ItemsDB();
        return sItemsDB;
    }

    public String listItems() {
        String r= "";
        for(HashMap.Entry<String, String> item : ItemsDB.entrySet())
            r= r+"\n Put " + item.getKey() + " in: " + item.getValue();
        return r;
    }

public String findCategory(String item){

        for (HashMap.Entry<String, String> i : ItemsDB.entrySet()){
            if (i.getKey().equals(item))
                return i.getValue();
        }
        return "Not found";

}

    // Will be used later
    public void addItem(String what, String where){
        ItemsDB.put(what, where);
    }

    public void fillItemsDB(Context context, String filename) throws IOException {
        BufferedReader reader = new BufferedReader( new InputStreamReader(context.getAssets().open(filename)) );
        String line = null;
        try {
            line = reader.readLine();
            while (line != null){
                String[] what_where = line.split(", ");
                String what = what_where[0];
                String where = what_where[1];
                ItemsDB.put(what, where);
            }
        } catch (IOException e) { }
    }

    public static void setContext(Context aContext) {
        context = aContext;
    }
}

File: GarbageV7-2025-Kotlin/app/src/main/java/dk/itu/garbagev7_2025_kotlin/ItemsDB.kt
package dk.itu.garbagev7_2025_kotlin

import android.content.Context
import androidx.lifecycle.Observer
import java.io.BufferedReader

class ItemsDB private constructor() {

    //handling files
    val items = ArrayList<Item>()

    //Singleton pattern
    companion object {
        @Volatile
        private var sItemsDB: ItemsDB? = null
        private var context: Context? = null

        fun setContext(ctx: Context) {
            context = ctx
        }

        fun getInstance(): ItemsDB {
            if(sItemsDB == null){
                sItemsDB = ItemsDB()
            }
            return sItemsDB!!
        }
    }

    //Constructor
    init { fillItemsDB(context!!,"garbage.txt") }

    //ADD
    fun addItem(what: String, where: String) {
        items.add(Item(what, where))
    }

    //DELETE
    fun removeItem(what: String, where: String) {
        items.removeIf { it.what == what && it.where == where }
    }


    fun size(): Int {
        return items.size
    }

    fun getWhere(what: String): String {
        for (item in items)
            if (item.what == what) return item.where
        return "not found"
    }

    fun isPresent(what: String): Boolean {
        return getWhere(what) != "not found"
    }

    private fun fillItemsDB(context: Context, fileName: String) {
        try {
            val bufferedReader = BufferedReader(context.assets.open(fileName).reader())
            bufferedReader.useLines {
                    lines -> lines.forEach {
                    line -> addItem(line.split(", ")[0].trim(), line.split(", ")[1].trim())
            }
            }
        }
        catch (e: Exception) { }
    }



}